CONFIG.md
Configuring MailHog
===================

You can configure MailHog using command line options or environment variables:

| Environment         | Command line  | Default         | Description
| ------------------- | ------------- | --------------- | -----------
| MH_CORS_ORIGIN      | -cors-origin  |                 | If set, a Access-Control-Allow-Origin header is returned for API endpoints
| MH_HOSTNAME         | -hostname     | mailhog.example | Hostname to use for EHLO/HELO and message IDs
| MH_HTTP_BIND_ADDR   | -httpbindaddr | 0.0.0.0:8025    | Interface and port for HTTP server to bind to
| MH_MONGO_COLLECTION | -mongocoll    | messages        | MongoDB collection name for message storage
| MH_MONGO_DB         | -mongodb      | mailhog         | MongoDB database name for message storage
| MH_MONGO_URI        | -mongouri     | 127.0.0.1:27017 | MongoDB host and port
| MH_SMTP_BIND_ADDR   | -smtpbindaddr | 0.0.0.0:1025    | Interface and port for SMTP server to bind to
| MH_STORAGE          | -storage      | memory          | Set message storage: memory / mongodb


docs/CONFIG.md
Configuring MailHog
===================

You can configure MailHog using command line options or environment variables:

| Environment         | Command line    | Default         | Description
| ------------------- | --------------- | --------------- | -----------
| MH_CORS_ORIGIN      | -cors-origin    |                 | If set, a Access-Control-Allow-Origin header is returned for API endpoints
| MH_HOSTNAME         | -hostname       | mailhog.example | Hostname to use for EHLO/HELO and message IDs
| MH_API_BIND_ADDR    | -api-bind-addr  | 0.0.0.0:8025    | Interface and port for HTTP UI server to bind to
| MH_UI_BIND_ADDR     | -ui-bind-addr   | 0.0.0.0:8025    | Interface and port for HTTP API server to bind to
| MH_MONGO_COLLECTION | -mongo-coll     | messages        | MongoDB collection name for message storage
| MH_MONGO_DB         | -mongo-db       | mailhog         | MongoDB database name for message storage
| MH_MONGO_URI        | -mongo-uri      | 127.0.0.1:27017 | MongoDB host and port
| MH_SMTP_BIND_ADDR   | -smtp-bind-addr | 0.0.0.0:1025    | Interface and port for SMTP server to bind to
| MH_STORAGE          | -storage        | memory          | Set message storage: memory / mongodb

#### Note on HTTP bind addresses

If `api-bind-addr` and `ui-bind-addr` are identical, a single listener will
be used allowing both to co-exist on one port.

The values must match in a string comparison. Resolving to the same host and
port combination isn't enough.


RELEASES.md
MailHog Releases
================

### [v0.08](https://github.com/mailhog/MailHog/releases/0.08)

- Extract SMTP protocol into isolated library
- Better protocol tests
- Add hooks for manipulating protocol behaviour
- Merge #14 - fix display of multipart messges, thanks @derwassi
- Merge #17 - fix API v1 docs, thanks @geerlingguy
- Fix #11 - add build documentation
- Fix #12 - support broken MAIL/RCPT syntax
- Fix #16 - add deployment documentation
- Fix #18 - better server-sent event support using [goose](https://github.com/ian-kent/goose)

### [v0.07](https://github.com/mailhog/MailHog/releases/tag/0.07)

- Fix #6 - Make SMTP verbs case-insensitive

### [v0.06](https://github.com/mailhog/MailHog/releases/tag/0.06)

- Fix #5 - Support leading tab in multiline headers

### [v0.05](https://github.com/mailhog/MailHog/releases/tag/0.05)

- Add #4 - UI support for RFC2047 encoded headers

### [v0.04](https://github.com/mailhog/MailHog/releases/tag/0.04)

* Configure from environment
* Include example Dockerfile
* Fix #1 - mismatched import path and repository name
* Fix #2 - possible panic with some MIME content
* Fix #3 - incorrect handling of RSET


### [v0.03](https://github.com/mailhog/MailHog/releases/tag/0.03)

* Download message in .eml format
* Cleaned up v1 API
* Web UI and API improvements
  * Fixed UI rendering bugs
  * Message search and matched/total message count
  * Message list resizing and scrolling  
  * EventSource support for message streaming
  * Better error handling and reporting
  * View/download individual MIME parts
  * Release messages to real SMTP servers
* Switch to [go-bindata](https://github.com/jteeuwen/go-bindata) for asset embedding

### [v0.02](https://github.com/mailhog/MailHog/releases/tag/0.02)

* Better support for ESMTP (RFC5321)
* Support for SMTP AUTH (RFC4954) and PIPELINING (RFC2920)
* Improved AJAX web interface to view messages (plain text, HTML or source)
* Improved HTTP API to list, retrieve and delete messages
* Multipart MIME support
* In-memory message storage
* MongoDB storage for message persistence

### [v0.01](https://github.com/mailhog/MailHog/releases/tag/0.01)

* Basic support for SMTP and HTTP servers
* Accepts SMTP messages
* Stores parsed messages in MongoDB
* Makes messages available via API
* has Bootstrap/AngularJS UI for viewing/deleting messages


MailHog-Server/config/config.go
package config

import (
	"flag"
	"log"

	"github.com/ian-kent/Go-MailHog/MailHog-Server/monkey"
	"github.com/ian-kent/Go-MailHog/data"
	"github.com/ian-kent/Go-MailHog/storage"
	"github.com/ian-kent/envconf"
)

func DefaultConfig() *Config {
	return &Config{
		SMTPBindAddr: "0.0.0.0:1025",
		HTTPBindAddr: "0.0.0.0:8025",
		Hostname:     "mailhog.example",
		MongoUri:     "127.0.0.1:27017",
		MongoDb:      "mailhog",
		MongoColl:    "messages",
		StorageType:  "memory",
		MessageChan:  make(chan *data.Message),
	}
}

type Config struct {
	SMTPBindAddr string
	HTTPBindAddr string
	Hostname     string
	MongoUri     string
	MongoDb      string
	MongoColl    string
	StorageType  string
	InviteJim    bool
	Storage      storage.Storage
	MessageChan  chan *data.Message
	Assets       func(asset string) ([]byte, error)
	Monkey       monkey.ChaosMonkey
}

var cfg = DefaultConfig()
var jim = &monkey.Jim{}

func Configure() *Config {
	switch cfg.StorageType {
	case "memory":
		log.Println("Using in-memory storage")
		cfg.Storage = storage.CreateInMemory()
	case "mongodb":
		log.Println("Using MongoDB message storage")
		s := storage.CreateMongoDB(cfg.MongoUri, cfg.MongoDb, cfg.MongoColl)
		if s == nil {
			log.Println("MongoDB storage unavailable, reverting to in-memory storage")
			cfg.Storage = storage.CreateInMemory()
		} else {
			log.Println("Connected to MongoDB")
			cfg.Storage = s
		}
	default:
		log.Fatalf("Invalid storage type %s", cfg.StorageType)
	}

	if cfg.InviteJim {
		jim.Configure(func(message string, args ...interface{}) {
			log.Printf(message, args...)
		})
		cfg.Monkey = jim
	}

	return cfg
}

func RegisterFlags() {
	flag.StringVar(&cfg.SMTPBindAddr, "smtpbindaddr", envconf.FromEnvP("MH_SMTP_BIND_ADDR", "0.0.0.0:1025").(string), "SMTP bind interface and port, e.g. 0.0.0.0:1025 or just :1025")
	flag.StringVar(&cfg.HTTPBindAddr, "httpbindaddr", envconf.FromEnvP("MH_HTTP_BIND_ADDR", "0.0.0.0:8025").(string), "HTTP bind interface and port, e.g. 0.0.0.0:8025 or just :8025")
	flag.StringVar(&cfg.Hostname, "hostname", envconf.FromEnvP("MH_HOSTNAME", "mailhog.example").(string), "Hostname for EHLO/HELO response, e.g. mailhog.example")
	flag.StringVar(&cfg.StorageType, "storage", envconf.FromEnvP("MH_STORAGE", "memory").(string), "Message storage: memory (default) or mongodb")
	flag.StringVar(&cfg.MongoUri, "mongouri", envconf.FromEnvP("MH_MONGO_URI", "127.0.0.1:27017").(string), "MongoDB URI, e.g. 127.0.0.1:27017")
	flag.StringVar(&cfg.MongoDb, "mongodb", envconf.FromEnvP("MH_MONGO_DB", "mailhog").(string), "MongoDB database, e.g. mailhog")
	flag.StringVar(&cfg.MongoColl, "mongocoll", envconf.FromEnvP("MH_MONGO_COLLECTION", "messages").(string), "MongoDB collection, e.g. messages")
	flag.BoolVar(&cfg.InviteJim, "invite-jim", envconf.FromEnvP("MH_INVITE_JIM", false).(bool), "Decide whether to invite Jim (beware, he causes trouble)")
	jim.RegisterFlags()
}
